src/modules/iotjs_module_mqtt.c

```diff
@@ -565,8 +587,175 @@ JS_FUNCTION(MqttHandle) {
     case SUBSCRIBE:
     case UNSUBSCRIBE:
     case PINGREQ:
+      return MQTT_ERR_UNALLOWED_PACKET;
+  }
+
+  return 0;
+}
+
+
+static void iotjs_mqtt_concat_buffers(iotjs_mqttclient_t *mqttclient,
+                                      iotjs_bufferwrap_t *buff_recv) {
+  char *tmp_buf = mqttclient->local_buffer;
+  mqttclient->local_buffer = (char *)malloc(
+      sizeof(char) * (buff_recv->length + mqttclient->buffer_length));
+  memcpy(mqttclient->local_buffer, tmp_buf, mqttclient->buffer_length);
+  memcpy(mqttclient->local_buffer + mqttclient->buffer_length,
+         buff_recv->buffer, buff_recv->length);
+  mqttclient->buffer_length += buff_recv->length;
+  free(tmp_buf);
+}
+
+
+static jerry_value_t iotjs_mqtt_handle_error(
+    iotjs_mqtt_packet_error_t error_code) {
+  switch (error_code) {
+    case MQTT_ERR_UNACCEPTABLE_PROTOCOL:
+      return JS_CREATE_ERROR(COMMON,
+                             "MQTT: Connection refused: unacceptable protocol");
+    case MQTT_ERR_BAD_IDENTIFIER:
+      return JS_CREATE_ERROR(COMMON,
+                             "MQTT: Connection refused: bad client identifier");
+    case MQTT_ERR_SERVER_UNAVIABLE:
+      return JS_CREATE_ERROR(COMMON,
+                             "MQTT: Connection refused: server unavailable");
+    case MQTT_ERR_BAD_CREDENTIALS:
       return JS_CREATE_ERROR(
-          COMMON, "MQTT: Unallowed packet has arrived to the client");
+          COMMON, "MQTT: Connection refused: bad username or password");
+    case MQTT_ERR_UNAUTHORISED:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Connection refused: unauthorised");
+    case MQTT_ERR_CORRUPTED_PACKET:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Corrupted packet");
+    case MQTT_ERR_UNALLOWED_PACKET:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Broker sent an unallowed packet");
+    case MQTT_ERR_SUBSCRIPTION_FAILED:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Subscription failed");
+    default:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Unknown error");
+  }
+}
+
+
+JS_FUNCTION(MqttReceive) {
+  DJS_CHECK_THIS();
+  DJS_CHECK_ARGS(2, object, object);
+
+  jerry_value_t jnat = JS_GET_ARG(0, object);
+
+  void *native_p;
+  JNativeInfoType *out_native_info;
+  bool has_p =
+      jerry_get_object_native_pointer(jnat, &native_p, &out_native_info);
+  if (!has_p || out_native_info != &mqttclient_native_info) {
+    return JS_CREATE_ERROR(COMMON, "MQTT native pointer not available");
+  }
+  iotjs_mqttclient_t *mqttclient = (iotjs_mqttclient_t *)native_p;
+
+  jerry_value_t jbuffer = JS_GET_ARG(1, object);
+  iotjs_bufferwrap_t *buff_recv = iotjs_bufferwrap_from_jbuffer(jbuffer);
+  IOTJS_ASSERT(buff_recv->length != 0);
+
+  // Concat the buffers if we previously needed to
+  if (mqttclient->must_concat) {
+    iotjs_mqtt_concat_buffers(mqttclient, buff_recv);
+    mqttclient->must_concat = false;
```

Comment:
I am not sure I understand the purpose of `must_concat`/ If the local buffer is not empty, we always need to do a concat first.